fastdtw/_fastdtw.pyx

```diff
@@ -0,0 +1,166 @@
+#!python
+#cython: boundscheck=False, cdivision=True, wraparound=False
+
+from __future__ import absolute_import, division
+
+from libcpp.vector cimport vector  # noqa
+from libcpp.map cimport map
+from libcpp.set cimport set
+from libcpp.utility cimport pair
+from cython.operator cimport dereference as deref
+from cython.operator cimport preincrement
+
+cdef struct CostEntry:
+    double cost
+    int prev_x_idx
+    int prev_y_idx
+
+
+def fastdtw(x, y, radius=1, dist=lambda a, b: abs(a - b)):
```

Comment:
Since there are a lot of recursive `fastdtw()` calls, it would make sense to expose this function in C (`cdef`). You can, instead, make a lightweight entry point function which ensures the arguments are correctly casted (i.e. into typed memoryviews) and then just passes the control flow to an internal recursive function (i.e. once you leave Python, be as much C as possible).
```cython
def fastdtw(x, y, radius=1, dist=lambda a, b: abs(a - b)):
    x = np.asanyarray(x, dtype=float)
    y = np.asanyarray(y, dtype=float)
    return _fastdtw(x, y, radius, dist)

cdef _fastdtw(double[:] x, double[:] y, int radius, dist):
    ...
```
Again, it's much easier and common with NumPy, but memoryviews can also be [constructed from cython.arrays](https://webcache.googleusercontent.com/search?strip=1&q=cache:http://docs.cython.org/src/userguide/memoryviews.html#cython-arrays).